### ISSUE TITLE
Performance of computation of recurrence matrices

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/RecurrenceAnalysis.jl#76]
Initially asked by @asinghvi17 , is it possible to parallelize or put on the GPU the calculation of Recurrence matrices? I have no knowledge on the GPU, so I will only talk about parallelization. The core function that does the number crunching is:

```julia
function _recurrence_matrix(xx::Dataset, yy::Dataset, ε, metric::Metric)
    x = xx.data
    y = yy.data
    rowvals = Vector{Int}()
    colvals = Vector{Int}()
    for j in 1:length(y)
        nzcol = 0
        for i in 1:length(x)
            @inbounds if evaluate(metric, x[i], y[j]) ≤ ε
                push!(rowvals, i)
                nzcol += 1
            end
        end
        append!(colvals, fill(j, (nzcol,)))
    end
    nzvals = fill(true, (length(rowvals),))
    return sparse(rowvals, colvals, nzvals, length(x), length(y))
end
```
You can see that the algorithm operates via `push!` instead of pre-allocating. It may be a tiny bit trick, but I don't think it should be impossible to parallelize this.

@heliosdrm do you have any take on this?

**Comment by Datseris** [comment]
By the way the line `    nzvals = fill(true, (length(rowvals),))` seems terribly inefficient. There has to be a way to do this smarter without calling `fill` everytime :P

**Comment by heliosdrm** [comment]
A very simple approach of parallelizing the calculation of large recurrence matrices would be to split the time series into smaller fragments, calculate the (cross-)recurrence matrices of different blocks separately (in parallel), and then join them.

E.g. if `x == [a; b]` and `y == [c; d]`, then `CrossRecurenceMatrix(x, y, ε)` should be the same as:

```julia
[CrossRecurrenceMatrix(a, c, ε)   CrossRecurrenceMatrix(a, d, ε);
 CrossRecurrenceMatrix(b, c, ε)   CrossRecurrenceMatrix(b, d, ε)]
```

But I don't know what criteria might be used to decide the appropriate size of the fragments for this to be efficient.

Regarding avoiding `fill`, you're right. Maybe it's worth exploring `SparseArrays` to see if there is some trick for this.

**Comment by Datseris** [comment]
oh ya! That's smart!

**Comment by Datseris** [comment]
I am changing the title, so we can use this issue to in-general discuss global performance of recurrence matrix computations, excluding specific cases like e.g. #60

**Comment by asinghvi17** [comment]
https://github.com/JuliaDynamics/RecurrenceAnalysis.jl/blob/ab07344cc4a6d43a9ab90577e09403b0f3392820/src/matrices.jl#L285

We could rewrite this loop to be parallel in 1.3.

**Comment by Datseris** [comment]
@asinghvi17 I haven't looked at all yet at 1.3 parallelization... Waiting for the stable release :P But you seem to have a better grasp than me at it, so I'll leave this to you.

BTW how do you manage to "cite" code??????? Is it sorcery?

**Comment by asinghvi17** [comment]
You can get a permalink to the code, by clicking on the line number in the Github interface.  If you paste that permalink into a comment, it's automatically formatted.
<img width="592" alt="Screen Shot 2019-09-20 at 8 50 15  96300AM" src="https://user-images.githubusercontent.com/32143268/65328316-0d7d6b00-db84-11e9-9d49-264eebbe2c52.png">

**Comment by asinghvi17** [comment]
> But I don't know what criteria might be used to decide the appropriate size of the fragments for this to be efficient.

Could we do some kind of performance assay on what the appropriate length is, and then use some form of automatic tiling?  We could also say that above a certain size, we just split into `JULIA_NUM_THREADS` segments (since that seems like the easiest solution).

**Comment by Datseris** [comment]
Yes, this is the best way forward I feel. We could set up a full pipeline using the `PkgBenchmark` package. We have done this in DynamicalBilliards.jl while it was after heavy development, and I found it very useful: https://github.com/JuliaDynamics/DynamicalBilliards.jl/blob/master/benchmark/benchmarks.jl

but it may also take a lot of effort to set up something like this. However `PkgBenchmark` seems to be under constant improvements and thus I think it will be overall worthy to use it.

**Comment by asinghvi17** [comment]
```julia
using Threads, RecurrenceAnalysis
function RecurrenceAnalysis._recurrence_matrix(xx::Dataset, yy::Dataset, ε, metric::Metric)
    x = xx.data
    y = yy.data
    rowvals = [Int[] for _ in Threads.nthreads()]
    colvals = [Int[] for _ in Threads.nthreads()]
    Threads.@threads for j in 1:length(y)
        threadn = Threads.threadid()
        nzcol = 0
        for i in 1:length(x)
            @inbounds if evaluate(metric, x[i], y[j]) ≤ ε
                push!(rowvals[threadn], i)
                nzcol += 1
            end
        end
        append!(colvals[threadn], fill(j, (nzcol,))) 
    end

    finalrows = vcat(rowvals...) # merge into one array
    finalcols = vcat(colvals...) # merge into one array
    nzvals = fill(true, (length(finalrows),))
    return sparse(finalrows, finalcols, nzvals, length(x), length(y))
end
```
is along the lines of what I'm thinking a threaded implementation could look like.  This isn't quite what @heliosdrm had in mind, but we'll have to test it (I haven't even run the code yet).

**Comment by asinghvi17** [comment]
So I have a working implementation:

If you start with `JULIA_NUM_THREADS` equal to whatever number of logical cores you have on your machine:
```julia
using Base.Threads, RecurrenceAnalysis, Distances, StaticArrays, BenchmarkTools
function RecurrenceAnalysis._recurrence_matrix(xx::Dataset, yy::Dataset, ε, metric::Metric)
                  x = xx.data
                  y = yy.data
                  rowvals = [Vector{Int}() for _ in 1:Threads.nthreads()]
                  colvals = [Vector{Int}() for _ in 1:Threads.nthreads()]
                  Threads.@threads for j in 1:length(y)
                      threadn = Threads.threadid()
                      nzcol = 0
                      for i in 1:length(x)
                          @inbounds if evaluate(metric, x[i], y[j]) ≤ ε
                              push!(rowvals[threadn], i)
                              nzcol += 1
                          end
                      end
                      append!(colvals[threadn], fill(j, (nzcol,)))
                  end

                  finalrows = vcat(rowvals...) # merge into one array
                  finalcols = vcat(colvals...) # merge into one array
                  nzvals = fill(true, (length(finalrows),))
                  return sparse(finalrows, finalcols, nzvals, length(x), length(y))
              end
```

Code to set up benchmark:
```julia
tr = sin.(LinRange(0, 10*pi, 30000) |> Dataset
```

Baseline benchmark (without parallelization):

```julia
BenchmarkTools.Trial:
  memory estimate:  688.58 MiB
  allocs estimate:  30067
  --------------
  minimum time:     2.364 s (3.44% GC)
  median time:      2.378 s (4.99% GC)
  mean time:        2.437 s (5.44% GC)
  maximum time:     2.568 s (7.70% GC)
  --------------
  samples:          3
  evals/sample:     1
```

With parallelization:

```julia
BenchmarkTools.Trial:
  memory estimate:  898.05 MiB
  allocs estimate:  30214
  --------------
  minimum time:     1.299 s (2.27% GC)
  median time:      1.414 s (10.89% GC)
  mean time:        1.412 s (10.03% GC)
  maximum time:     1.522 s (15.07% GC)
  --------------
  samples:          4
  evals/sample:     1
```

**Comment by Datseris** [comment]
COOL.

Just one thing: I would assume that there is a size of dataset that this fails (e.g. for very small datasets I would imagine that the unparallel version is faster). Can we confirm this? Do we need a heuristic in choosing this size? 

In my eyes, I think it may be worth considering having two different functions: one parallelized, one not. I have had trouble with parallelization on serial clusters and this may help.

**Comment by pucicu** [comment]
I agree. It would be good to have a test for different but very large sizes of time series.

Perhaps using GPU for calculation could be a further option for the near future?

**Comment by Datseris** [comment]
Ah, now I realize that #22 also belongs in this discussion.

**Comment by asinghvi17** [comment]
I'm reopening this issue, because there's one approach which has been left out of this discussion, which is computing only the lower triangle of the matrix.  I tried a naive method in the parallel scheme today, and it actually _increased_ the runtime compared to the parallel version (understandably, since it ends up placing a majority of the burden on one thread).  

I suspect we will need some kind of partitioning scheme, which can partition the triangle more or less equally between threads.  I guess the cases we could specialize on would be 4, 6, and 8 threads?  Beyond that, I guess you can just make your own...
